src/go/mongolib/fingerprinter/fingerprinter.go

```diff
@@ -271,3 +269,80 @@ func deduplicate(s []string) (r []string) {
 
 	return r
 }
+
+// GetQueryFieldD returns the correct field to build the fingerprint, based on the operation.
+func GetQueryFieldD(doc proto.SystemProfile) (primitive.M, error) {
+	// Proper way to detect if protocol used is "op_msg" or "op_command"
+	// would be to look at "doc.Protocol" field,
+	// however MongoDB 3.0 doesn't have that field
+	// so we need to detect protocol by looking at actual data.
+	query := doc.Query
+	if len(doc.Command) > 0 {
+		query = doc.Command
+		if doc.Op == "update" || doc.Op == "remove" {
+			if squery, ok := query.Map()["q"]; ok {
+				switch v := squery.(type) {
```

Comment:
can we use `asMap` here?

Reply:
There no more check if `query.Map()["q"]` exists or not

Reply:
@percona-csalguero

Reply:
No. It will return an error if it not exists.